# <center>Homework Assignment 1: Data Cleaning and Wrangling</center>
## <center>27th August 2024</center>
#### <center>Syed Asghar Abbas Zaidi 07201 in CS/CE 457/464-L1</center>
#### <center>sz07201@st.habib.edu.pk</center>


In [7]:
import pandas as pd
data = pd.read_csv('580SurveyCleanup2.csv')

In [9]:
data.head()

Unnamed: 0       Unnamed: 1 Unnamed: 2          Unnamed: 3   Unnamed: 4  \
0        No.             Time     Status  Duration (Seconds)  Access Code   
1          1  10/3/2023 17:49  Completed               11646          NaN   
2          2  10/3/2023 17:49  Completed                 113          NaN   
3          3  10/3/2023 17:49  Completed                 572          NaN   
4          4  10/3/2023 17:50  Completed                 365          NaN   

      Unnamed: 5      Unnamed: 6     Unnamed: 7 Unnamed: 8  \
0  Email Address      IP Address        Country     Region   
1            NaN  100.36.167.226  United States   Virginia   
2            NaN   98.169.26.134  United States   Virginia   
3            NaN    98.169.52.23  United States   Virginia   
4            NaN  69.255.182.253  United States   Virginia   

               Unnamed: 9  ...             Q12.1         Q12.2  \
0  City / Town / District  ...  Some familiarity  Average user   
1                 Fairfax  ...                 1           NaN   
2                 Fairfax  ...               NaN           NaN   
3          Greenway Downs  ...               NaN             1   
4             Gainesville  ...                 1           NaN   

                       Q12.3          Q12.4  Q13 Q13.1  Q14 Q14.1     Q14.2  \
0  Frequent use for projects  Fluent/expert  Yes    No  Yes    No  Not sure   
1                        NaN            NaN    1   NaN    1   NaN       NaN   
2                        NaN            NaN    1   NaN  NaN     1       NaN   
3                        NaN            NaN  NaN     1  NaN   NaN         1   
4                        NaN            NaN  NaN     1    1   NaN       NaN   

                                               Q13.2  
0  What is your goal for learning about data anal...  
1                                                NaN  
2                                                NaN  
3  As part of my work, I have access to a lot of ...  
4                                                NaN  

[5 rows x 51 columns]

## Formatting AIT-580 Section

In [30]:
#data_exp = data.iloc[:,[0,1]]
# data_exp = data.iloc[10, :]

# data_exp

In [36]:
data_AIT = data.iloc[:,[12,13]]
data_AIT.head()

Q1  Q1.1
0  2.0   4.0
1  NaN   1.0
2  1.0   NaN
3  1.0   NaN
4  1.0   NaN

In [38]:
data_AIT.columns = data_AIT.iloc[0].astype(str)
data_AIT.head()

0  2.0  4.0
0  2.0  4.0
1  NaN  1.0
2  1.0  NaN
3  1.0  NaN
4  1.0  NaN

In [40]:
data_AIT = data_AIT[1:]
data_AIT.head()

0  2.0  4.0
1  NaN  1.0
2  1.0  NaN
3  1.0  NaN
4  1.0  NaN
5  1.0  NaN

In [42]:
data_AIT = data_AIT.loc[:, '2.0':'4.0'].replace(1, pd.Series(data_AIT.columns, data_AIT.columns)).fillna('')
data_AIT.head()

0  2.0  4.0
1       4.0
2  2.0     
3  2.0     
4  2.0     
5  2.0

In [453]:
data_AIT2 = pd.DataFrame(data_AIT.iloc[:,0] + data_AIT.iloc[:,1]) #concatenate both Q1 columns into one final column in a new dataframe
data_AIT2.columns = ['AIT-580 Section'] #Renaming the column from 0 to Q1

In [455]:
data_AIT2.head()

AIT-580 Section
1             4.0
2             2.0
3             2.0
4             2.0
5             2.0

## Formatting Gender Section

In [44]:
data_gender = data.iloc[:,14:18]
data_gender.columns = data_gender.iloc[0].astype(str)
data_gender = data_gender[1:]
data_gender = data_gender.apply(pd.to_numeric, errors='coerce') #converting 1s to actual integers warna replace will not work
data_gender = data_gender.replace(1, pd.Series(data_gender.columns, data_gender.columns)).fillna('')
data_gender.head()
# print(data_gender.dtypes) #Checking datatype,

0  Male  Female NonBinary  Prefer not to answer
1        Female                                
2        Female                                
3  Male                                        
4                          Prefer not to answer
5  Male

In [46]:
data_gender2 = pd.DataFrame(data_gender.iloc[:,0] + data_gender.iloc[:,1] + data_gender.iloc[:,2] + data_gender.iloc[:,3]) #concatenate
data_gender2.columns = ['Gender?'] #Renaming the column from 0 to Q1
data_gender2.head()

Gender?
1                Female
2                Female
3                  Male
4  Prefer not to answer
5                  Male

## Formatting Laptop Used Q8

In [55]:
data_laptop = data.iloc[:,23:27]
data_laptop.loc[0] = data_laptop.loc[0].fillna('Other')
data_laptop.columns = data_laptop.iloc[0].astype(str)
data_laptop = data_laptop[1:]
data_laptop.iloc[:,:3] = data_laptop.iloc[:,:3].apply(pd.to_numeric, errors='coerce') #converting 1s to actual integers warna replace will not work
data_laptop.head()


0 Microsoft/Windows Apple/MacBook Other Other
1               NaN           1.0   NaN   NaN
2               NaN           1.0   NaN   NaN
3               NaN           1.0   NaN   NaN
4               1.0           NaN   NaN   NaN
5               NaN           2.0   NaN   NaN

In [57]:
data_laptop = data_laptop.replace(1, pd.Series(data_laptop.columns, data_laptop.columns)).fillna('')
data_laptop = data_laptop.replace(0, pd.Series(data_laptop.columns, data_laptop.columns)).fillna('') #why is there 0 
data_laptop = data_laptop.replace(2, pd.Series(data_laptop.columns, data_laptop.columns)).fillna('') #why is there 2
data_laptop.head()

0  Microsoft/Windows  Apple/MacBook Other Other
1                     Apple/MacBook            
2                     Apple/MacBook            
3                     Apple/MacBook            
4  Microsoft/Windows                           
5                     Apple/MacBook

In [59]:
# data_laptop.head()
# print(data_gender.dtypes) #Checking datatype,
data_laptop = pd.concat([data_laptop.iloc[:,:2],data_laptop.iloc[:,3]],axis=1)
data_laptop2 = pd.DataFrame(data_laptop.iloc[:,0] + data_laptop.iloc[:,1] + data_laptop.iloc[:,2]) #concatenate
data_laptop2.columns = ['Laptop Type']
data_laptop2.head()

Laptop Type
1      Apple/MacBook
2      Apple/MacBook
3      Apple/MacBook
4  Microsoft/Windows
5      Apple/MacBook

## Employment Status While Attending Mason University

In [469]:
df1 = data.iloc[:,27:30]
df1.columns = df1.iloc[0].astype(str)
df1 = df1[1:]
df1 = df1.apply(pd.to_numeric, errors='coerce') #converting 1s to actual integers warna replace will not work
df1 = df1.replace(1, pd.Series(df1.columns, df1.columns)).fillna('')
data_df1 = pd.DataFrame(df1.iloc[:,0] + df1.iloc[:,1] + df1.iloc[:,2]) #concatenate
data_df1.columns = ['Employment Status while attending Mason']
data_df1.head()

Employment Status while attending Mason
1       Not Working while attending Mason
2              Working, but not Full Time
3                          Yes, Full Time
4       Not Working while attending Mason
5                          Yes, Full Time

## Python-Skill Level (Q10)

In [472]:
df2 = data.iloc[:,30:35]
df2.columns = df2.iloc[0].astype(str)
df2 = df2[1:]
df2 = df2.apply(pd.to_numeric, errors='coerce') #converting 1s to actual integers warna replace will not work
df2 = df2.replace(1, pd.Series(df2.columns, df2.columns)).fillna('')
data_df2 = pd.DataFrame(df2.iloc[:,0] + df2.iloc[:,1] + df2.iloc[:,2] + df2.iloc[:,3] + df2.iloc[:,4]) #concatenate
data_df2.columns = ['Python-Skill Level']
data_df2.head()

data_df2.iloc[:, 0] = data_df2.iloc[:, 0].replace('Little/none', 1)
data_df2.iloc[:, 0] = data_df2.iloc[:, 0].replace('Some familiarity', 2)
data_df2.iloc[:, 0] = data_df2.iloc[:, 0].replace('Average user', 3)
data_df2.iloc[:, 0] = data_df2.iloc[:, 0].replace('Frequent use for projects', 4)
data_df2.iloc[:, 0] = data_df2.iloc[:, 0].replace('Fluent/expert', 5)

data_df2.head()

Python-Skill Level
1                  2
2                  3
3                  2
4                  3
5                  2

## R-skill Level (Q11)

In [475]:
df3 = data.iloc[:,35:40]
df3.columns = df3.iloc[0].astype(str)
df3 = df3[1:]
df3 = df3.apply(pd.to_numeric, errors='coerce') #converting 1s to actual integers warna replace will not work
df3 = df3.replace(1, pd.Series(df3.columns, df3.columns)).fillna('')
data_df3 = pd.DataFrame(df3.iloc[:,0] + df3.iloc[:,1] + df3.iloc[:,2] + df3.iloc[:,3] + df3.iloc[:,4]) #concatenate
data_df3.columns = ['R-Skill Level']

data_df3.iloc[:, 0] = data_df3.iloc[:, 0].replace('Little/none', 1)
data_df3.iloc[:, 0] = data_df3.iloc[:, 0].replace('Some familiarity', 2)
data_df3.iloc[:, 0] = data_df3.iloc[:, 0].replace('Average user', 3)
data_df3.iloc[:, 0] = data_df3.iloc[:, 0].replace('Frequent use for projects', 4)
data_df3.iloc[:, 0] = data_df3.iloc[:, 0].replace('Fluent/expert', 5)

data_df3.head()

R-Skill Level
1             1
2             3
3             1
4             4
5             2

## SQL-skill Level (Q12)

In [478]:
df4 = data.iloc[:,40:45]
df4.columns = df4.iloc[0].astype(str)
df4 = df4[1:]
df4 = df4.apply(pd.to_numeric, errors='coerce') #converting 1s to actual integers warna replace will not work
df4 = df4.replace(1, pd.Series(df4.columns, df4.columns)).fillna('')
data_df4 = pd.DataFrame(df4.iloc[:,0] + df4.iloc[:,1] + df4.iloc[:,2] + df4.iloc[:,3] + df4.iloc[:,4]) #concatenate
data_df4.columns = ['SQL Skill-Level']

data_df4.iloc[:, 0] = data_df4.iloc[:, 0].replace('Little/none', 1)
data_df4.iloc[:, 0] = data_df4.iloc[:, 0].replace('Some familiarity', 2)
data_df4.iloc[:, 0] = data_df4.iloc[:, 0].replace('Average user', 3)
data_df4.iloc[:, 0] = data_df4.iloc[:, 0].replace('Frequent use for projects', 4)
data_df4.iloc[:, 0] = data_df4.iloc[:, 0].replace('Fluent/expert', 5)

data_df4.head()

SQL Skill-Level
1               2
2               1
3               3
4               2
5               3

## Github Used (Q13)

In [481]:
df5 = data.iloc[:,45:47]
df5.columns = df5.iloc[0].astype(str)
df5 = df5[1:]
df5 = df5.apply(pd.to_numeric, errors='coerce') #converting 1s to actual integers warna replace will not work
# Ensure 'Yes' and 'No' are column names in df5
df5 = df5.replace(1, pd.Series(df5.columns, df5.columns)).fillna('')
data_df5 = pd.DataFrame(df5.iloc[:,0] + df5.iloc[:,1]) #concatenate
data_df5.columns = ['Github Used?']
data_df5.head()

Github Used?
1          Yes
2          Yes
3           No
4           No
5           No

## Will you pursue your career in Data Science/Data Analytics field? 

In [484]:
df6 = data.iloc[:,47:50]
df6.columns = df6.iloc[0].astype(str)
df6 = df6[1:]
df6 = df6.apply(pd.to_numeric, errors='coerce') #converting 1s to actual integers warna replace will not work
df6 = df6.replace(1, pd.Series(df6.columns, df6.columns)).fillna('')
data_df6 = pd.DataFrame(df6.iloc[:,0] + df6.iloc[:,1] + df6.iloc[:,2]) #concatenate
data_df6.columns = ['Plans to pursue career in Data Science/Analytics?']
data_df6.head()

Plans to pursue career in Data Science/Analytics?
1                                               Yes
2                                                No
3                                          Not sure
4                                               Yes
5                                               Yes

## Goal for learning Data Analynatics

In [487]:
df7 = data.iloc[:,50:51]
df7.columns = df7.iloc[0].astype(str)
df7 = df7.fillna('')
data_df7 = df7[1:]
data_df7.head()

0 What is your goal for learning about data analytics?
1                                                     
2                                                     
3  As part of my work, I have access to a lot of ...  
4                                                     
5                                              Learn

## Merging and final Formatting!

### Removing the **.No**, **Time taken** to fill the form, **Status. of the form**,  **Access Code** and **Email Address** column is being removed cause it has no data

**Access Code and Email Address field removal:** The survey dataset shows no recorded entries for the access code and email address field, with all values being NaN (Not available). This could indicate that either the access code was not a required field or it was not captured correctly during data collection. As a result, no analysis and insights can be performed on access codes in this dataset, that's why I am removing it as part of my formatting of the data.
Serial Numbers although useful for our eyes isn't necessarily useful and might take needless storage space.

**Serial Numbers Removal:** Removing serial numbers from a CSV file during data cleaning is important because they are often arbitrary and irrelevant to the analysis, potentially introducing noise or misleading insights if treated as meaningful data. By eliminating them, you streamline the dataset, reduce its size, and improve the focus on relevant variables, ensuring cleaner and more accurate results.

**Why am I keeping IP addresses:** I am not removing the IP addresses as they can potentially let us know if there were _duplicates_ responses, and possibly tell us about the general geographical location but aside of that, they don't have much use.

**Time to fill survey removal:** Time to fill the form is irrelevant for my survey unless I want to reconduct the survey and want to gauge the "respodent's engagement". Longer time might indicate that certain questions were pretty confusing or too complex. Very short times could indicate that they skimmed through the survey. Although this survey isn't of that kind, it could potentially showcase which people from where are more familiar with the questions being asked. Again, I feel like this data is of not particular use here so I am removing it.

**Why keep Longitude and Latitude**: Although we have already asked the country and region, keeping this information may help us cross-verify and understand certain outliers. Keeping the latitude and longitude of your respondents can provide valuable geographic insights, such as identifying regional patterns, trends, or differences in responses.

**Why am I removing columns here?** Data formatting often involves structuring the dataset to make it more manageable and relevant for analysis, which includes dropping columns that don't contribute meaningful information. Part 2 (data cleaning) would then focus on refining the remaining data, handling missing values, and standardizing content.

In [491]:
#Handling extra stuff
df_e = data.iloc[:,6:12]
df_e.columns = df_e.iloc[0].astype(str)
df_e = df_e.fillna('')
df_e = df_e[1:]
df_e.head()

0      IP Address        Country    Region City / Town / District  Latitude  \
1  100.36.167.226  United States  Virginia                Fairfax  -77.2891   
2   98.169.26.134  United States  Virginia                Fairfax   -77.263   
3    98.169.52.23  United States  Virginia         Greenway Downs  -77.2331   
4  69.255.182.253  United States  Virginia            Gainesville  -77.6308   
5  138.88.108.230  United States                                    -97.822   

0 Longitude  
1   38.8209  
2   38.8605  
3   38.9645  
4   38.8175  
5    37.751

In [493]:
#Handling extra stuff
df_e2 = data.iloc[:,18:23]
df_e2.columns = df_e2.iloc[0].astype(str)
df_e2 = df_e2[1:]
df_e2.head()

0 Age (years) Height (Inches) Country of Citizenship   Undergraduate Degree  \
1          23              63                 Taiwan              economics   
2          24              59                  India               Finance    
3          40              70                  India  Computer Applications   
4          26              66            South Korea    Business Management   
5          30              70                     US                    Yes   

0 Expected Graduation date from Mason MS program?  
1                                      12/31/2021  
2                                        5/1/2021  
3                                        6/1/2022  
4                                      12/31/2021  
5                                        1/1/2021

In [495]:
df_merge1 = pd.concat([df_e, data_AIT2,data_gender2],axis=1)
df_merge2 = pd.concat([df_e2, data_laptop2, data_df1,data_df2,data_df3,data_df4,data_df5,data_df6,data_df7],axis=1)
df_merge_main = pd.concat([df_merge1, df_merge2],axis=1)
df_merge_main

IP Address        Country    Region City / Town / District  Latitude  \
1    100.36.167.226  United States  Virginia                Fairfax  -77.2891   
2     98.169.26.134  United States  Virginia                Fairfax   -77.263   
3      98.169.52.23  United States  Virginia         Greenway Downs  -77.2331   
4    69.255.182.253  United States  Virginia            Gainesville  -77.6308   
5    138.88.108.230  United States                                    -97.822   
6     108.56.134.80  United States  Virginia                 Reston  -77.3489   
7    108.48.175.220  United States  Virginia              Arlington  -77.0987   
8     173.67.199.85  United States  Virginia                Ashburn   -77.539   
9      74.96.114.12  United States  Virginia             Woodbridge  -77.3197   
10    107.77.241.32  United States     Texas                         -96.8217   
11   138.88.182.245  United States  Maryland            Takoma Park   -77.004   
12     72.83.246.60  United States  Virginia                Fairfax   -77.263   
13    96.241.43.105  United States  Virginia              Chantilly  -77.4512   
14    71.191.72.223  United States  Virginia                  Burke  -77.2866   
15    73.134.53.227  United States  Maryland            Hyattsville  -76.8696   
16     73.86.219.52  United States  Maryland              Baltimore  -76.6355   
17   69.138.215.127  United States  Virginia              Arlington   -77.113   
18    72.196.231.70  United States  Virginia            Springfield  -77.2333   
19     100.15.54.17  United States  Virginia             Woodbridge  -77.3197   
20   184.170.83.239  United States  Virginia               Rustburg  -79.0985   
21   100.36.243.102  United States  Virginia                Fairfax  -77.3375   
22    108.56.134.80  United States  Virginia                 Reston  -77.3489   
23    73.163.163.82  United States  Maryland              Baltimore   -76.557   
24    138.88.223.87  United States                                    -97.822   
25    100.36.50.114  United States                                    -97.822   
26    108.51.59.195  United States  Virginia                  Burke  -77.2866   
27    72.83.246.219  United States  Virginia                Fairfax   -77.263   
28   100.36.118.186  United States  Virginia                  Burke  -77.2866   
29    172.58.187.98  United States  Maryland              Baltimore  -76.6949   
30   72.196.230.201  United States  Virginia            Springfield  -77.2333   
31   108.18.111.204  United States  Virginia                 Lorton   -77.206   
32    71.191.72.223  United States  Virginia                  Burke  -77.2866   
33     184.191.72.8  United States  Virginia           Falls Church  -77.1425   
34  141.156.160.243  United States                                    -97.822   
35    71.206.51.121  United States  Maryland              Frederick  -77.2758   
36  129.174.252.234  United States  Virginia                Fairfax  -77.3375   

   Longitude AIT-580 Section               Gender? Age (years)  \
1    38.8209             4.0                Female          23   
2    38.8605             2.0                Female          24   
3    38.9645             2.0                  Male          40   
4    38.8175             2.0  Prefer not to answer          26   
5     37.751             2.0                  Male          30   
6    38.9311             2.0                Female          25   
7      38.86             2.0             NonBinary          24   
8     39.018             2.0                Female          30   
9     38.678             2.0                  Male          24   
10   32.7787             4.0                  Male          26   
11   38.9827             2.0                Female          24   
12   38.8605             4.0                Female          25   
13   38.9036             2.0                  Male          25   
14   38.7851             2.0                Female          23   
15   38.9544         

As we can observe, we reduced the initial survey data from 51 columns to 21 columns successfully

## Saving to "Formatted_Output" CSV file

In [499]:
output_filename = "formatted_output.csv"

# Save the DataFrame to a CSV file with a header
df_merge_main.to_csv(output_filename, index=False)

print(f"DataFrame saved to {output_filename}")

DataFrame saved to formatted_output.csv


# Part 2 (Cleaning the Data)

In [502]:
data_p2 = pd.read_csv('formatted_output.csv')
import numpy as np

## Cleaning Regions

In [505]:
data_p2['Region'].value_counts()

Region
Virginia    25
Maryland     6
Texas        1
Name: count, dtype: int64

In [507]:
region_null = data_p2['Region'].isnull().sum()
print(f"Although there are {region_null} empty spaces, I won't be filling them using the mode. Although there is a higher likelihood that they would belong to that region, we can't say for sure.")


Although there are 4 empty spaces, I won't be filling them using the mode. Although there is a higher likelihood that they would belong to that region, we can't say for sure.


In [509]:
data_p2['Region'] = data_p2['Region'].fillna('Unknown') #Just filling in "Unknown" to let us know that we don't know their region.
data_p2['Region'].head()

0    Virginia
1    Virginia
2    Virginia
3    Virginia
4     Unknown
Name: Region, dtype: object

## Cleaning Age
To address the issues with the 'Age' column, first, I am replacing incorrect values like '6l' and '7O' with valid numbers ('6' and '70', respectively). Next, convert the column to numeric data types to ensure all entries are valid numbers. Finally, fill any remaining NaN values with the median height.

The **median was used** instead of the mean because the median is less sensitive to outliers and extreme values, which makes it a more robust measure of central tendency for skewed or non-normally distributed data. If the age data includes unusual values or errors, like incorrect entries, the median provides a more accurate central value that is less influenced by these anomalies compared to the mean.
It is generally a safer bet

In [512]:
data_p2['Age (years)'].head()

0    23
1    24
2    40
3    26
4    30
Name: Age (years), dtype: object

In [514]:
age_clean = data_p2['Age (years)'].str.replace(' ', '', regex=True)
age_clean = pd.to_numeric(age_clean, errors='coerce')
age_clean.isnull().sum() #No null values so no need to put any missing values

0

**Now I am handling the "1993" case here properly.** There is a valid chance that someone might mistakenly put down their year of birth instead of simply putting down their age.

In such scenerios, this helper function detect this and will subtract from the year given from the current year, to find out their real age. 

In [517]:
from datetime import datetime
def calculate_age(value):
    # Check if the value is between 1900 and 2100
    if 1900 <= value <= datetime.now().year:
        current_year = datetime.now().year
        age = current_year - value
        return age
    elif value > datetime.now().year: #This is just in case if someone gives a huge value
        return None
    else:
        return value  # Or handle the case when value is out of range

data_p2['Age (years)'] = age_clean.apply(calculate_age)
data_p2['Age (years)']

0     23
1     24
2     40
3     26
4     30
5     25
6     24
7     30
8     24
9     26
10    24
11    25
12    25
13    23
14    31
15    26
16    40
17    23
18    25
19    22
20    23
21    25
22    22
23    23
24    27
25    28
26    23
27    35
28    27
29    25
30    55
31    23
32    31
33    23
34    24
35    28
Name: Age (years), dtype: int64

## Cleaning height

To address the issues with the 'Height (Inches)' column, first, I am replacing incorrect values like '6l' and '7O' with valid numbers ('6' and '70', respectively). Next, convert the column to numeric data types to ensure all entries are valid numbers. Finally, fill any remaining NaN values with the median height.

The **median was used** instead of the mean because the median is less sensitive to outliers and extreme values, which makes it a more robust measure of central tendency for a potentially skewed or non-normally distributed data. 

In [521]:
# Define a function to clean the 'Height (Inches)' column
def clean_height(value):
    try:
        # Attempt to convert value to a numeric type
        value = str(value).replace('O', '0').replace('l', '1')  # Handle common misinterpretations
        return pd.to_numeric(value, errors='coerce')  # Convert to float, coercing errors to NaN
    except:
        return np.nan

# Apply the cleaning function to the 'Height (Inches)' column
data_p2['Height (Inches)'] = data_p2['Height (Inches)'].apply(clean_height)
# Calculate the median
median_height = data_p2['Height (Inches)'].median()

# Fill NaNs with the median and reassign
data_p2['Height (Inches)'] = data_p2['Height (Inches)'].fillna(median_height)
data_p2['Height (Inches)'].head()

0    63.0
1    59.0
2    70.0
3    66.0
4    70.0
Name: Height (Inches), dtype: float64

## Cleaning Country of Citizenship

In [524]:
# Standardize country names
data_p2['Country of Citizenship'] = data_p2['Country of Citizenship'].replace({
    'US': 'United States',
    'USA': 'United States',
})

# Capitalize the first letter of each country name
data_p2['Country of Citizenship'] = data_p2['Country of Citizenship'].str.title()
data_p2['Country of Citizenship'] = data_p2['Country of Citizenship'].str.rstrip() #removing trailing whitespaces

#Filling it "Unknown" for null 
data_p2['Country of Citizenship'] = data_p2['Country of Citizenship'].fillna('Unknown')
data_p2['Country of Citizenship']

0                       Taiwan
1                        India
2                        India
3                  South Korea
4                United States
5                      Unknown
6                     Pakistan
7                      Unknown
8                        China
9                   Uzbekistan
10               United States
11                       India
12                     Romania
13                       India
14                       China
15                  Uzbekistan
16               United States
17                       India
18                    Ethiopia
19               United States
20               United States
21                       India
22               United States
23                     Nigeria
24               United States
25                      Taiwan
26                       India
27                     Nigeria
28                       India
29                       India
30               United States
31                       India
32    Un

For handling missing data in the Undergraduate Degree column, NaN values were replaced with 'Unknown' to ensure all entries are non-null and to maintain consistency. The entry 'Yes' was also renamed to 'Unknown' because it did not fit the context of an undergraduate degree and was likely a data entry error or an ambiguous response.

In [527]:
Degree = data_p2['Undergraduate Degree'] 

In [529]:
# Clean the column
Degree = data_p2['Undergraduate Degree'].str.strip()  # Remove leading/trailing whitespace
Degree = Degree.str.title()  # Capitalize first letter of each word
Degree = Degree.fillna('Unknown')

idx = Degree.str.contains('Yes')
Degree.loc[idx] = 'Unknown'

Degree = Degree.replace({
    'Economic': 'Economics',
    'Bit': 'Information Technology',
})

Degree.head()

0                Economics
1                  Finance
2    Computer Applications
3      Business Management
4                  Unknown
Name: Undergraduate Degree, dtype: object

In [531]:
data_p2['Undergraduate Degree'] = Degree
data_p2['Undergraduate Degree'] .head()

0                Economics
1                  Finance
2    Computer Applications
3      Business Management
4                  Unknown
Name: Undergraduate Degree, dtype: object

## Cleaning Dates

In [534]:
dates = data_p2['Expected Graduation date from Mason MS program?']
dates.head()

0    12/31/2021
1      5/1/2021
2      6/1/2022
3    12/31/2021
4      1/1/2021
Name: Expected Graduation date from Mason MS program?, dtype: object

In [536]:
data_p2['Expected Graduation date from Mason MS program?'] = pd.to_datetime(data_p2['Expected Graduation date from Mason MS program?'], format='%m/%d/%Y')
# format to mm/yyyy
data_p2['Expected Graduation date from Mason MS program?'] = data_p2['Expected Graduation date from Mason MS program?'].dt.strftime('%m/%Y')

In [538]:
dates = data_p2['Expected Graduation date from Mason MS program?']
data_p2['Expected Graduation date from Mason MS program?'].isnull().sum()

3

In [540]:
data_p2['Expected Graduation date from Mason MS program?'].head()

0    12/2021
1    05/2021
2    06/2022
3    12/2021
4    01/2021
Name: Expected Graduation date from Mason MS program?, dtype: object

In [542]:
data_p2['Expected Graduation date from Mason MS program?'].value_counts()

Expected Graduation date from Mason MS program?
12/2021    14
05/2021     5
05/2022     5
01/2021     2
12/2022     2
06/2022     1
07/2022     1
08/2021     1
04/2022     1
05/2023     1
Name: count, dtype: int64

We will be filling the missing data with "mode". The reason for choosing mode is because there's a subsequent number of people graduating in 12/2021 so there is very high proability that the missing data would be similar to this as well. If not then our analysis won't deviate much. Keeping it null for the sake of trasnparency and to have no biases is also a reasonable approach but here it won't affect our analysis much

In [545]:
data_p2['Expected Graduation date from Mason MS program?'] = data_p2['Expected Graduation date from Mason MS program?'].fillna(data_p2['Expected Graduation date from Mason MS program?'].mode()[0])

In [547]:
data_p2['Expected Graduation date from Mason MS program?'].head()

0    12/2021
1    05/2021
2    06/2022
3    12/2021
4    01/2021
Name: Expected Graduation date from Mason MS program?, dtype: object

To have our data be more readable, I am removing the dates, and only keeping the month and year of graduation.

## Cleaning the Laptop Column

In [551]:
data_p2['Laptop Type'].value_counts()

Laptop Type
Microsoft/Windows                 21
Apple/MacBook                      8
Microsoft/WindowsApple/MacBook     3
Acer/Windows                       1
Name: count, dtype: int64

In [553]:
data_p2['Laptop Type'].isnull().sum()

3

In [557]:
# replace WindowsMacBook with Both
data_p2['Laptop Type'] = data_p2['Laptop Type'].replace('Microsoft/WindowsApple/MacBook', 'Both')

As we can observe from the count above, that Microsoft/Windows is clearly far greater in numbers and are the winner. We can use "mode" to fill in rest of the values

In [567]:
# replace '' with NaN
data_p2['Laptop Type'] = data_p2['Laptop Type'].replace('', None)
laptop_mode = data_p2['Laptop Type'].mode()[0]
data_p2['Laptop Type'] = data_p2['Laptop Type'].fillna(laptop_mode)
data_p2['Laptop Type'].value_counts()

Laptop Type
Microsoft/Windows    24
Apple/MacBook         8
Both                  3
Acer/Windows          1
Name: count, dtype: int64

## Cleaning the Employment column

In [576]:
data_p2['Employment Status while attending Mason'].value_counts()

Employment Status while attending Mason
Not Working while attending Mason    17
Yes, Full Time                       11
Working, but not Full Time            5
Name: count, dtype: int64

In [579]:
data_p2['Employment Status while attending Mason'].isnull().sum()

3

In [582]:
data_p2['Employment Status while attending Mason'] = data_p2['Employment Status while attending Mason'].replace('', None)

In [590]:
mode_job = data_p2['Employment Status while attending Mason'].mode()[0]
mode_job

'Not Working while attending Mason'

In [596]:
data_p2['Employment Status while attending Mason'] = data_p2['Employment Status while attending Mason'].fillna(mode_job)
data_p2['Employment Status while attending Mason'].isnull().sum()

0

## Cleaning Python Skill Level

In [601]:
data_p2['Python-Skill Level'].value_counts()

Python-Skill Level
3.0    15
2.0    12
1.0     4
4.0     2
Name: count, dtype: int64

In [603]:
data_p2['Python-Skill Level'].isnull().sum()

3

Median is generally the best choice for filling in missing skill level values, as it is both representative and resistant to outliers. However, if one particular skill level (e.g., 3) is overwhelmingly common, mode could also be a strong option.

In [610]:
data_p2['Python-Skill Level'] = data_p2['Python-Skill Level'].replace('', None)
m_python = data_p2['Python-Skill Level'].median()
data_p2['Python-Skill Level'] = data_p2['Python-Skill Level'].fillna(m_python).astype(int)
data_p2['Python-Skill Level'].value_counts()

Python-Skill Level
3    18
2    12
1     4
4     2
Name: count, dtype: int64

## Cleaning R and SQL Skill Level
We are using median for the same reason we used it on Python skill level above!

In [615]:
data_p2['R-Skill Level'].value_counts()

R-Skill Level
3.0    9
1.0    8
2.0    8
4.0    7
5.0    1
Name: count, dtype: int64

In [617]:
data_p2['R-Skill Level'].isnull().sum()

3

In [619]:
data_p2['R-Skill Level'] = data_p2['R-Skill Level'].replace('', None)
m_R = data_p2['R-Skill Level'].median()
data_p2['R-Skill Level'] = data_p2['R-Skill Level'].fillna(m_R).astype(int)
data_p2['R-Skill Level'].value_counts()

R-Skill Level
3    12
1     8
2     8
4     7
5     1
Name: count, dtype: int64

In [630]:
data_p2['R-Skill Level'].isnull().sum()

0

In [623]:
data_p2['SQL Skill-Level'].value_counts()

SQL Skill-Level
3.0    15
2.0    10
1.0     5
4.0     3
Name: count, dtype: int64

In [625]:
data_p2['SQL Skill-Level'].isnull().sum()

3

In [627]:
data_p2['SQL Skill-Level'] = data_p2['SQL Skill-Level'].replace('', None)
m_SQL = data_p2['SQL Skill-Level'].median()
data_p2['SQL Skill-Level'] = data_p2['SQL Skill-Level'].fillna(m_SQL).astype(int)
data_p2['SQL Skill-Level'].value_counts()

SQL Skill-Level
3    18
2    10
1     5
4     3
Name: count, dtype: int64

In [632]:
data_p2['SQL Skill-Level'].isnull().sum()

0

## Cleaning City
Again, just like region, you could use "mode" and then insert the most common district there (which should be Fairfax here) as they have a very higher probability from being there, but just in case, I don't want to misrepresent the data so I won't be doing that and instead will keep it purposely "unknown". 

In [636]:
data_p2['City / Town / District'] = data_p2['City / Town / District'].fillna('Unknown')
data_p2['City / Town / District']

0            Fairfax
1            Fairfax
2     Greenway Downs
3        Gainesville
4            Unknown
5             Reston
6          Arlington
7            Ashburn
8         Woodbridge
9            Unknown
10       Takoma Park
11           Fairfax
12         Chantilly
13             Burke
14       Hyattsville
15         Baltimore
16         Arlington
17       Springfield
18        Woodbridge
19          Rustburg
20           Fairfax
21            Reston
22         Baltimore
23           Unknown
24           Unknown
25             Burke
26           Fairfax
27             Burke
28         Baltimore
29       Springfield
30            Lorton
31             Burke
32      Falls Church
33           Unknown
34         Frederick
35           Fairfax
Name: City / Town / District, dtype: object

In [645]:
data_p2

IP Address        Country    Region City / Town / District  Latitude  \
0    100.36.167.226  United States  Virginia                Fairfax  -77.2891   
1     98.169.26.134  United States  Virginia                Fairfax  -77.2630   
2      98.169.52.23  United States  Virginia         Greenway Downs  -77.2331   
3    69.255.182.253  United States  Virginia            Gainesville  -77.6308   
4    138.88.108.230  United States   Unknown                Unknown  -97.8220   
5     108.56.134.80  United States  Virginia                 Reston  -77.3489   
6    108.48.175.220  United States  Virginia              Arlington  -77.0987   
7     173.67.199.85  United States  Virginia                Ashburn  -77.5390   
8      74.96.114.12  United States  Virginia             Woodbridge  -77.3197   
9     107.77.241.32  United States     Texas                Unknown  -96.8217   
10   138.88.182.245  United States  Maryland            Takoma Park  -77.0040   
11     72.83.246.60  United States  Virginia                Fairfax  -77.2630   
12    96.241.43.105  United States  Virginia              Chantilly  -77.4512   
13    71.191.72.223  United States  Virginia                  Burke  -77.2866   
14    73.134.53.227  United States  Maryland            Hyattsville  -76.8696   
15     73.86.219.52  United States  Maryland              Baltimore  -76.6355   
16   69.138.215.127  United States  Virginia              Arlington  -77.1130   
17    72.196.231.70  United States  Virginia            Springfield  -77.2333   
18     100.15.54.17  United States  Virginia             Woodbridge  -77.3197   
19   184.170.83.239  United States  Virginia               Rustburg  -79.0985   
20   100.36.243.102  United States  Virginia                Fairfax  -77.3375   
21    108.56.134.80  United States  Virginia                 Reston  -77.3489   
22    73.163.163.82  United States  Maryland              Baltimore  -76.5570   
23    138.88.223.87  United States   Unknown                Unknown  -97.8220   
24    100.36.50.114  United States   Unknown                Unknown  -97.8220   
25    108.51.59.195  United States  Virginia                  Burke  -77.2866   
26    72.83.246.219  United States  Virginia                Fairfax  -77.2630   
27   100.36.118.186  United States  Virginia                  Burke  -77.2866   
28    172.58.187.98  United States  Maryland              Baltimore  -76.6949   
29   72.196.230.201  United States  Virginia            Springfield  -77.2333   
30   108.18.111.204  United States  Virginia                 Lorton  -77.2060   
31    71.191.72.223  United States  Virginia                  Burke  -77.2866   
32     184.191.72.8  United States  Virginia           Falls Church  -77.1425   
33  141.156.160.243  United States   Unknown                Unknown  -97.8220   
34    71.206.51.121  United States  Maryland              Frederick  -77.2758   
35  129.174.252.234  United States  Virginia                Fairfax  -77.3375   

    Longitude  AIT-580 Section               Gender?  Age (years)  \
0     38.8209              4.0                Female           23   
1     38.8605              2.0                Female           24   
2     38.9645              2.0                  Male           40   
3     38.8175              2.0  Prefer not to answer           26   
4     37.7510              2.0                  Male           30   
5     38.9311              2.0                Female           25   
6     38.8600              2.0             NonBinary           24   
7     39.0180              2.0                Female           30   
8     38.6780              2.0                  Male           24   
9     32.7787              4.0                  Male           26   
10    38.9827              2.0                Female           24   
11    38.8605              4.0                Female           25   
12    38.9036              2.0                  Male           25   
13    38.7851              2.0               

In [638]:
output_filename = "cleaned_output.csv"

# Save the DataFrame to a CSV file with a header
data_p2.to_csv(output_filename, index=False)

print(f"DataFrame saved to {output_filename}")

DataFrame saved to cleaned_output.csv


# Part 3 (Statistical Summarization!)

In [640]:
data_p3 = pd.read_csv('cleaned_output.csv')
import numpy as np

In [677]:
k = data_p3
k = k.drop(columns='What is your goal for learning about data analytics?')
k2 = k.describe(include='all')
k2.iloc[:,4:18]

Latitude  Longitude  AIT-580 Section Gender?  Age (years)  \
count   36.000000  36.000000        36.000000      36    36.000000   
unique        NaN        NaN              NaN       4          NaN   
top           NaN        NaN              NaN    Male          NaN   
freq          NaN        NaN              NaN      18          NaN   
mean   -80.093406  38.562475         2.888889     NaN    27.166667   
std      7.151399   1.101628         1.007905     NaN     6.518107   
min    -97.822000  32.778700         2.000000     NaN    22.000000   
25%    -77.374475  38.730625         2.000000     NaN    23.000000   
50%    -77.286600  38.835700         2.000000     NaN    25.000000   
75%    -77.226325  38.931100         4.000000     NaN    28.000000   
max    -76.557000  39.462400         4.000000     NaN    55.000000   

        Height (Inches) Country of Citizenship Undergraduate Degree  \
count         36.000000                     36                   36   
unique              NaN                     13                   25   
top                 NaN                  India            Economics   
freq                NaN                     10                    5   
mean          65.755556                    NaN                  NaN   
std           11.169459                    NaN                  NaN   
min            5.200000                    NaN                  NaN   
25%           66.000000                    NaN                  NaN   
50%           67.000000                    NaN                  NaN   
75%           70.000000                    NaN                  NaN   
max           75.000000                    NaN                  NaN   

       Expected Graduation date from Mason MS program?        Laptop Type  \
count                                               36                 36   
unique                                              10                  4   
top                                            12/2021  Microsoft/Windows   
freq                                                17                 24   
mean                                               NaN                NaN   
std                                                NaN                NaN   
min                                                NaN                NaN   
25%                                                NaN                NaN   
50%                                                NaN                NaN   
75%                                                NaN                NaN   
max                                                NaN                NaN   

       Employment Status while attending Mason  Python-Skill Level  \
count                                       36           36.000000   
unique                                       3                 NaN   
top          Not Working while attending Mason                 NaN   
freq                                        20                 NaN   
mean                                       NaN            2.500000   
std                                        NaN            0.774597   
min                                        NaN            1.000000   
25%                                        NaN            2.000000   
50%                                        NaN            3.000000   
75%                                        NaN            3.000000   
max                                        NaN            4.000000   

        R-Skill Level  SQL Skill-Level  
count       36.000000        36.000000  
unique            NaN              NaN  
top               NaN              NaN  
freq              NaN              NaN  
mean         2.583333         2.527778  
std          1.130739         0.844685  
min          1.000000         1.000000  
25%          2.000000         2.000000  
50%          3.000000         3.000000  
75%          3.000000         3.000000  
max          5.000000         4.000000

From the table above, it can be referred that most students took "Economics"
### **Latitude**

The latitude data indicates that respondents are spread across various geographical locations, with an average latitude of -80.09. The values range from -97.822 to -76.557, which suggests that participants are likely located across different regions, potentially spanning from central to eastern parts of the United States. The standard deviation of 7.15 further confirms that the respondents are geographically diverse.

### **Longitude**

Longitude data shows less variability compared to latitude, with a mean of 38.56 and a standard deviation of 1.10. This suggests that respondents are more clustered in terms of longitude, likely situated along a particular meridian. The data range from 32.779 to 39.462 indicates that participants might be concentrated within a specific state or region.

In [686]:
k2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 11 entries, count to max
Data columns (total 20 columns):
 #   Column                                             Non-Null Count  Dtype  
---  ------                                             --------------  -----  
 0   IP Address                                         4 non-null      object 
 1   Country                                            4 non-null      object 
 2   Region                                             4 non-null      object 
 3   City / Town / District                             4 non-null      object 
 4   Latitude                                           8 non-null      float64
 5   Longitude                                          8 non-null      float64
 6   AIT-580 Section                                    8 non-null      float64
 7   Gender?                                            4 non-null      object 
 8   Age (years)                                        8 non-null      float64
 9   Height (Inch

In [703]:
data_p2['AIT-580 Section'].value_counts()

AIT-580 Section
2.0    20
4.0    16
Name: count, dtype: int64

Both sections were almost equally represented, with Section 004 receiving a slightly higher number of responses.

### Q2 - Gender

In [713]:
data_p2['Gender?'].value_counts()

Gender?
Male                    18
Female                  16
Prefer not to answer     1
NonBinary                1
Name: count, dtype: int64

This suggests that there was nearly equal gender representation among the students who took the class, assuming that all enrolled students also completed the survey.

### Q3 - age

In [721]:
data_p2['Age (years)'].value_counts()

Age (years)
23    8
25    6
24    5
26    3
40    2
30    2
31    2
22    2
27    2
28    2
35    1
55    1
Name: count, dtype: int64

In [724]:
data_p2['Age (years)'].describe()

count    36.000000
mean     27.166667
std       6.518107
min      22.000000
25%      23.000000
50%      25.000000
75%      28.000000
max      55.000000
Name: Age (years), dtype: float64

The age distribution of the students shows a mean of 27.17 years and a standard deviation of 6.52 years, reflecting considerable variability. Ages range from 22 to 55 years, with a median of 25 years, indicating that half of the students are younger than 25. The interquartile range spans from 23 to 28 years, suggesting that most students are within this age range. The maximum age being notably higher than the 75th percentile points to the presence of older outliers.

### Q4 height

In [734]:
data_p2['Height (Inches)'].value_counts()

Height (Inches)
67.0    6
70.0    5
66.0    5
68.0    3
71.0    3
72.0    2
65.0    1
56.0    1
64.0    1
61.0    1
75.0    1
63.0    1
74.0    1
60.0    1
59.0    1
69.0    1
5.2     1
73.0    1
Name: count, dtype: int64

In [736]:
data_p2['Height (Inches)'].describe()

count    36.000000
mean     65.755556
std      11.169459
min       5.200000
25%      66.000000
50%      67.000000
75%      70.000000
max      75.000000
Name: Height (Inches), dtype: float64


The height data indicates that respondents have an average height of 65.76 inches, approximately 5 feet 6 inches. Most participants' heights fall within a typical range for adults, with the 25th to 75th percentiles spanning from 66 to 70 inches. However, the dataset includes an outlier with a minimum value of 5.2 inches, which likely represents an error. Excluding this, the height measurements align well with general population norms.

### Q5 Country of citizenship

In [746]:
data_p2['Country of Citizenship'].value_counts()

Country of Citizenship
India                       10
United States                9
Uzbekistan                   3
Taiwan                       2
Unknown                      2
China                        2
Nigeria                      2
South Korea                  1
Pakistan                     1
Romania                      1
Ethiopia                     1
United States ; Pakistan     1
Vietnam                      1
Name: count, dtype: int64

The class exhibited significant diversity, with approximately two-thirds of the students hailing from outside the US, primarily from Asia, and notably a large number from India. This suggests that the course attracted a wide range of individuals, likely due to the strong career prospects it offers in a lucrative field. The prominent presence of Asian students reflects a broader trend where individuals from this region seek high-paying job opportunities abroad, and pursuing such courses helps facilitate their career advancement.

### Undergraduate Degree Q6

In [750]:
data_p2['Undergraduate Degree'].value_counts()

Undergraduate Degree
Economics                   5
Unknown                     3
Finance                     3
Computer Science            2
Economics With Finance      2
Bachelor Of Engineering     2
Criminology Law And Soci    1
Bs Accounting               1
B.Tech In Cs                1
Quantity Surveying          1
Ece                         1
Civil Engineering           1
Systems Engineering         1
Mathematics & Statistics    1
Accounting And Isom         1
Statistics                  1
Engineering                 1
Management                  1
Cse                         1
Biology                     1
Computer Engineering        1
Information Technology      1
Business Management         1
Computer Applications       1
Psychology                  1
Name: count, dtype: int64

The distribution of undergraduate degrees among the students shows a diverse academic background. The most common degrees are Economics, Unknown, and Finance, each with 3 to 5 students. Other notable fields include Computer Science, Economics with Finance, and various engineering disciplines, each represented by 2 students. There is also a range of unique degrees such as Criminology Law and Sociology, Quantity Surveying, and Psychology, each with just one student. This variety indicates a broad spectrum of academic interests and backgrounds among the students.

### Graduation Date (Q7)

In [756]:
data_p2['Expected Graduation date from Mason MS program?'].value_counts()

Expected Graduation date from Mason MS program?
12/2021    17
05/2021     5
05/2022     5
01/2021     2
12/2022     2
06/2022     1
07/2022     1
08/2021     1
04/2022     1
05/2023     1
Name: count, dtype: int64

The distribution of expected graduation dates from the Mason MS program reveals a concentration of students graduating in December 2021, with 17 students indicating this date. May 2021 and May 2022 are also notable, each with 5 students. Other dates such as January 2021, December 2022, and several others have fewer students, ranging from 1 to 2. This pattern shows a significant cluster of graduations around the end of 2021, with a gradual spread into subsequent years.

In [759]:
data_p2['Expected Graduation date from Mason MS program?'].describe()

count          36
unique         10
top       12/2021
freq           17
Name: Expected Graduation date from Mason MS program?, dtype: object

### Q8 Laptop 

In [765]:
data_p2['Laptop Type'].describe()

count                    36
unique                    4
top       Microsoft/Windows
freq                     24
Name: Laptop Type, dtype: object

In [ ]:
data_p2['Laptop Type'].value_counts()

The distribution of laptop types among the students shows a clear preference for Microsoft/Windows laptops, with 24 out of 36 students using them. Apple/MacBook users are the second most common, with 8 students. A smaller number of students use both types of laptops (3) or an Acer/Windows laptop (1). This suggests that Microsoft/Windows is the dominant choice among the students, while Apple/MacBook and other types are less frequently used.

In [785]:
k2.iloc[:,14:18]

Employment Status while attending Mason  Python-Skill Level  \
count                                       36           36.000000   
unique                                       3                 NaN   
top          Not Working while attending Mason                 NaN   
freq                                        20                 NaN   
mean                                       NaN            2.500000   
std                                        NaN            0.774597   
min                                        NaN            1.000000   
25%                                        NaN            2.000000   
50%                                        NaN            3.000000   
75%                                        NaN            3.000000   
max                                        NaN            4.000000   

        R-Skill Level  SQL Skill-Level  
count       36.000000        36.000000  
unique            NaN              NaN  
top               NaN              NaN  
freq              NaN              NaN  
mean         2.583333         2.527778  
std          1.130739         0.844685  
min          1.000000         1.000000  
25%          2.000000         2.000000  
50%          3.000000         3.000000  
75%          3.000000         3.000000  
max          5.000000         4.000000

### Employment 

In [789]:
data_p2['Employment Status while attending Mason'].describe()

count                                    36
unique                                    3
top       Not Working while attending Mason
freq                                     20
Name: Employment Status while attending Mason, dtype: object

The employment status data for students while attending Mason shows that 20 out of 36 students reported not working during their studies. The most common status is "Not Working while attending Mason," reflecting a significant majority. The dataset also includes two other employment statuses, though they are less frequent. This indicates that a substantial portion of the students chose to focus on their studies without concurrent employment

### Python Q10

In [795]:
data_p2['Python-Skill Level'].describe()

count    36.000000
mean      2.500000
std       0.774597
min       1.000000
25%       2.000000
50%       3.000000
75%       3.000000
max       4.000000
Name: Python-Skill Level, dtype: float64

Given that a skill level of 1 is the least and 5 is the most, the analysis of Python skill levels shows a mean of 2.5 and a standard deviation of 0.77, reflecting moderate variability in skill levels. The range extends from a minimum of 1 to a maximum of 4, with a median skill level of 3, indicating that at least half of the students rate their skills as 3 or higher. The interquartile range, from the 25th percentile (2) to the 75th percentile (3), suggests that the majority of students fall within this intermediate skill range. **This indicates a concentration of skill levels around the middle of the scale, with a tendency towards modest proficiency.**


### R skill Q11

In [802]:
data_p2['R-Skill Level'].describe()

count    36.000000
mean      2.583333
std       1.130739
min       1.000000
25%       2.000000
50%       3.000000
75%       3.000000
max       5.000000
Name: R-Skill Level, dtype: float64

The analysis of R skill levels among the students shows a mean of 2.58, with a standard deviation of 1.13, indicating significant variability in skill levels. The skill levels range from a minimum of 1 to a maximum of 5. The median skill level is 3, meaning that half of the students rate their R skills as 3 or higher. The interquartile range, from the 25th percentile (2) to the 75th percentile (3), suggests that most students' skill levels fall within this intermediate range. **This distribution highlights a central tendency towards moderate proficiency with considerable variation in skill levels.**

### SQL skill Q12

In [804]:
data_p2['SQL Skill-Level'].describe()

count    36.000000
mean      2.527778
std       0.844685
min       1.000000
25%       2.000000
50%       3.000000
75%       3.000000
max       4.000000
Name: SQL Skill-Level, dtype: float64

The analysis of SQL skill levels among the students reveals a mean of 2.53 and a standard deviation of 0.84, indicating moderate variability in skill levels. The range spans from a minimum of 1 to a maximum of 4. The median skill level is 3, suggesting that at least half of the students rate their SQL skills as 3 or higher. The interquartile range, from the 25th percentile (2) to the 75th percentile (3), indicates that most students' skill levels fall within this intermediate range. This distribution reflects a concentration around moderate proficiency in SQL, with some variability among students.

In [818]:
k2.iloc[:,18:20]

Github Used? Plans to pursue career in Data Science/Analytics?
count            36                                                36
unique            2                                                 3
top              No                                               Yes
freq             24                                                17
mean            NaN                                               NaN
std             NaN                                               NaN
min             NaN                                               NaN
25%             NaN                                               NaN
50%             NaN                                               NaN
75%             NaN                                               NaN
max             NaN                                               NaN

### Github Q13

In [820]:
data_p2['Github Used?'].describe()

count     36
unique     2
top       No
freq      24
Name: Github Used?, dtype: object

The data on GitHub usage among the students shows that 24 out of 36 students do not use GitHub, making "No" the most common response. Only 12 students indicated that they use GitHub. This suggests a notable majority of students do not utilize GitHub, highlighting a potential area for increasing awareness or training on version control tools among the student group.

### Plans to pursue career in Data Science/Analytics? Q14

In [824]:
data_p2['Plans to pursue career in Data Science/Analytics?'].describe()

count      36
unique      3
top       Yes
freq       17
Name: Plans to pursue career in Data Science/Analytics?, dtype: object

The data on career plans reveals that 17 out of 36 students plan to pursue a career in Data Science or Analytics, making "Yes" the most common response. The remaining students are split between other career paths or undecided, with 19 students indicating they do not plan to pursue this field or are unsure. This suggests a significant interest in Data Science and Analytics among the students, reflecting the growing appeal of these fields.

# Extra!
Initially I had Duration as a column and even analysized it before ultiamtely deciding to remove it for reasons provided in first part. But still here is my analysis for that 
### **Duration (Seconds)**

The data reveals that respondents took an average of 7571.03 seconds (approximately 2.1 hours) to complete the survey. However, the duration varies widely, with some completing it in as little as 40 seconds, while others took over 35 hours (128,253 seconds). The large standard deviation of 28292.03 seconds highlights significant variability in the time taken, possibly due to differing levels of engagement, distractions, or technical issues encountered during the survey process.